### PR TITLE
Support Online FP8 & NF4 Weight Quantization

### PR DESCRIPTION
**Pull request by Uxito-Ada**
## Description

Add online FP8 & NF4 weight-only quantization for DiT pipelines with torchao and bitsandbytes backend: 
- introduce `QuantType.TORCHAO_FP8` and  `QuantType.BNB_NF4`
- Qwen-Image FP8 and NF4 examples on H100

## Motivation

Online weight-only quantization is necessary especially for models not natively providing quantized checkpoints:
- FP8 can reduce memory size to some extend and keep high quality at the same.
- NF4 can sharply reduce memory size further, which is useful for low-memory GPUs, and also keep high quality compared with lossy solution e.g. torchao INT4.
Detailed data and comparison can be found in the following.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance improvement
- [x] Code refactoring
- [x] Documentation update
- [ ] Other (please describe):

## Changes Made

- Add `QuantType.TORCHAO_FP8`, `QuantType.BNB_NF4`, `QuantKernelBackend.TORCHAO` and `QuantKernelBackend.BITSANDBYTES`
- Add `telefuser.ops.torchao_fp8_linear` and `telefuser.ops.bnb_nf4_linear`
- Added `qwen_image_t2i_torchao_fp8_h100.py ` and `qwen_image_t2i_bnb_nf4_h100.py `, as well as according documents

## Benchmark Setup

- GPU: NVIDIA H100 80GB HBM3
- Model: Qwen-Image-2512
- Prompt: `A cat playing piano`
- Resolution: 1328x1328, 928x1664
- Steps: 8, 16, 32
- Loop: 2 warmup, 5 repeats
- Dtype:
  - BF16 (baseline)
  - Qwen-Image Lightning FP8
  - BNB NF4
  - TorchAO INT4
  - TorchAO FP8
  - llmcompressor FP8
- Metrics:
  - Latency/Throughput
  - Memory
  - Quality

## Performance Results

image

image

- Though the performance of torchao FP8 is slightly weaker than Lightning FP8, it is still necessary for models without FP8 checkpoint.
- llmcompressor is not chosen is because that it does not save any memory: it is found that llmcompressor does not convert much Qwen-Image's DiT BF16 linears to FP8. Although llmcompressor declares that it has W8A8, it shows much slower throughput, even worse than the other low bits.
- BNB NF4 is chosen rather than Tochao NF4 is because that it shows much better image quality.

### Image Quality Comparison
qwen_image_quality_comparison_table

## Testing

- [x] Unit tests pass (`pytest tests/`)
- [x] Manual testing performed
- [x] Benchmarks added/updated (if applicable)

Test commands:

```bash
pip install torchao

PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True \
python examples/qwen_image/qwen_image_t2i_telefuser_fp8_h100.py \
  --prompt "A cat playing piano" \
  --aspect_ratio 1:1 \
  --num-inference-steps 16 \
  --seed 42 \
  --output qwen_image_fp8.png
```

```bash
pip install -U bitsandbytes

# set the below cuda versions according to your environment
export BNB_CUDA_VERSION=128
export CUDA_HOME=/usr/local/cuda-12.8
export LD_LIBRARY_PATH=/usr/local/cuda-12.8/lib64:${LD_LIBRARY_PATH:-}

PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True \
python examples/qwen_image/qwen_image_t2i_telefuser_nf4_h100.py   \
    --prompt "A cat playing piano"   \
    --aspect_ratio 1:1   \
    --num-inference-steps 16   \
    --seed 42   \
    --output qwen_image_nf4.png
```

## Checklist

- [x] Code follows the project's coding standards (`ruff`)
- [x] Pre-commit hooks pass (`pre-commit run --all-files`)
- [x] All tests pass (`pytest tests/`)
- [x] New tests added for new functionality
- [x] Documentation updated (example and PR notes)
- [x] Commit messages are clear and descriptive
- [x] PR title follows the convention: `[TYPE] Brief description`

## GPU Architecture Support

- [ ] SM80 (Ampere, Ada Lovelace)
- [x] SM90 (Hopper H100)
- [ ] SM100+ (Blackwell)

### CODE REVIEW

**Comment by Uxito-Ada** [comment]
LLMCompressor's FP8 is an implementation of **smoothquant** W8A8, which shows great quality, but it degrades too much for latency, because it calculates activation statistics at runtime profiling, and has no advantage in terms of memory @lzx1413

File: telefuser/core/base_stage.py [hunk 1]
```diff
@@ -142,6 +142,27 @@ def parallel_models(self) -> None:
         """Apply parallelization to models. Override in subclass."""
         pass
 
+    def apply_model_quantization(self, model: Any, runtime_config: ModelRuntimeConfig | None = None) -> None:
```
Comment (lzx1413) [inline telefuser/core/base_stage.py]: Some stages (e.g., denoising in WAN 2.2) use multiple models; apply this in the stage's init function.

**Comment by lzx1413** [comment]
> LLMCompressor's FP8 is an implementation of **smoothquant** W8A8, which shows great quality, but it degrades too much for latency, because it calculates activation statistics at runtime profiling, and has no advantage in terms of memory @lzx1413

We can adopt an offline calibration Post-Training Quantization (PTQ) process to create the quantized model.

File: examples/qwen_image/qwen_image_t2i_telefuser_fp8_h100.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    text_encoder_paths = [os.path.join(model_root, p) for p in PPL_CONFIG["text_encoder_path_list"]]
+    tokenizer_path = os.path.join(model_root, PPL_CONFIG["tokenizer_path"])
+
+    quant_config = QuantConfig(
+        enabled=True,
+        quant_type=QuantType.TORCHAO_FP8,
+        kernel_backend=QuantKernelBackend.TORCHAO,
+    )
+    mm = ModuleManager(torch_dtype=torch.bfloat16, device="cpu")
+    mm.load_model(dit_paths, device="cuda", torch_dtype=torch.bfloat16, quant_config=quant_config)
```
Comment (Uxito-Ada) [inline examples/qwen_image/qwen_image_t2i_telefuser_fp8_h100.py:80]: Both torchao and bitsandbytes needs to quant with tensors on GPU, so module_manager directly loads to device cuda.

Only quantize the DiT, as claimed by papers that quantization on encoder degrades the token inputs, and on vae decoder degrades the output image quality.

Quantized 840 linears of Qwen-Image-2512's DiT.

The above is the same as for NF4.

@lzx1413

File: telefuser/core/module_manager.py [hunk 3]
```diff
@@ -78,12 +74,22 @@ def load_model_from_single_file(
         if not low_cpu_mem_usage:
             model_state_dict = {k: v.to("cpu").clone() for k, v in model_state_dict.items()}
 
-        # Load weights and move to target device/dtype
-        model.load_state_dict(model_state_dict, strict=strict, assign=True)
+        # Load weights and move to target device/dtype first. Runtime quantization
```
Comment (Uxito-Ada) [inline telefuser/core/module_manager.py:77]: Removed the `apply_model_quantization` which previously converts the dit in the pipelines/dit_denoising files.

By contrast, uniformly call `enable_quant` at `module_manager`.

The difference between them with `torch.float8_e4m3fn` is that, both torchao and bitsandbytes require on-GPU convert, so the model needs to be moved to GPU before calling `enable_quant`.

@lzx1413

**Comment by Uxito-Ada** [comment]
> > LLMCompressor's FP8 is an implementation of **smoothquant** W8A8, which shows great quality, but it degrades too much for latency, because it calculates activation statistics at runtime profiling, and has no advantage in terms of memory @lzx1413
> 
> We can adopt an offline calibration Post-Training Quantization (PTQ) process to create the quantized model.

I think we can do this in another PR :)

File: docs/quantization/telefuser_fp8_deployment.md
Comment (lzx1413) [inline docs/quantization/telefuser_fp8_deployment.md:1]: Merge usage commands into the example folder's README. Create `quantization.md` in the `docs` folder, with both English and Chinese versions.

Comment (Uxito-Ada) [reply docs/quantization/telefuser_fp8_deployment.md:1]: refactored

**Comment by lzx1413** [comment]
LGTM
